community/kernel/src/main/java/org/neo4j/kernel/impl/store/id/FreeIdKeeper.java

```diff
@@ -116,155 +148,165 @@ else if ( defraggedIdCount > 0 && canReadMoreIdBatches() )
-                }
-            }
-            writeBuffer.flip();
-            while ( writeBuffer.hasRemaining() )
+            long id = freeIds.remove( 0 );
+            if ( id == NO_RESULT )
             {
-                channel.write( writeBuffer );
+                continue;
             }
-            if ( aggressiveReuse )
+            writeBuffer.putLong( id );
+            if ( writeBuffer.position() == writeBuffer.capacity() )
             {
-                maxReadPosition = channel.size();
+                writeBuffer.flip();
+                channel.writeAll( writeBuffer );
+                writeBuffer.clear();
             }
         }
-        catch ( IOException e )
+        writeBuffer.flip();
+        if ( writeBuffer.hasRemaining() )
         {
-            throw new UnderlyingStorageException(
-                    "Unable to write defragged id " + " batch", e );
+            channel.writeAll( writeBuffer );
         }
+        return channel.position();
     }
 
-    private void positionChannel( long newPosition ) throws IOException
+    /*
+     * Writes both freeIds and readFromDisk lists to disk and truncates the channel to size.
+     * It forces but does not close the channel.
+     */
+    @Override
+    public void close() throws IOException
     {
-        if ( newPosition < lowWatermarkForChannelPosition )
+        ByteBuffer writeBuffer = ByteBuffer.allocate( batchSize * ID_ENTRY_SIZE );
+        flushFreeIds( writeBuffer );
+        freeIds.addAll( readFromDisk );
+        flushFreeIds( writeBuffer );
+        if ( !aggressiveMode )
         {
-            throw new IllegalStateException( String.format( "%d is less than the lowest position (%d) this id keeper " +
-                    "can go", newPosition, lowWatermarkForChannelPosition ) );
+            compact( writeBuffer );
         }
-        channel.position( newPosition );
+        channel.force( false );
     }
 
-    private void defragReusableIdsInFile( ByteBuffer writeBuffer ) throws IOException
+    /**
+     * Compacts away the gap which will form in non-aggressive mode
+     * when batches are read in from disk.
+     *
+     * The gap will contain already used IDs so it is important to remove it
+     * on a clean shutdown. The freed IDs will not be reused after an
+     * unclean shutdown, as guaranteed by the external user.
+     */
+    private void compact( ByteBuffer writeBuffer ) throws IOException
     {
-        if ( readPosition > lowWatermarkForChannelPosition )
+        if ( prePosition == stackPosition )
         {
-            long writePosition = lowWatermarkForChannelPosition;
-            long position = Math.min( readPosition, maxReadPosition );
-            int bytesRead;
-            do
+            // there is no compaction to be done
+            return;
+        }
+
+        long writePosition = stackPosition;
+        long readPosition = prePosition;
+        int nBytes;
+        do
```

Comment:
not quite sure how it works for cases when write position is higher then read position

Reply:
That cannot happen because this is only for the regular mode. The stack can only grow and the reader can only consume down towards 0. The initial state is that they are equal and if they still are when closing/compacting then there is an early return. In all other cases the batches will be moved from prePosition to stackPosition.

Reply:
yeah now i see, would you mind to add that additional assertion still into that method that `stackPosition` should be less than `prePosition`

Reply:
Definitely!